lib/events/athena/consumer.go

```diff
@@ -0,0 +1,460 @@
+			if ctx.Err() != nil {
+				c.Debug("Batcher has been stopped")
+				return
+			}
+			c.Errorf("Batcher single run failed: %v", err)
+		}
+	}
+}
+
+// singleBatch creates single batch of events. It waits either up to BatchMaxInterval
+// or BatchMaxItems while reading events from queue. Batch is sent to s3 as
+// parquet file and at the end events are deleted from queue.
+func (c *consumer) singleBatch(ctx context.Context) error {
+	start := time.Now()
+	var size int
+	// TODO(tobiaszheller): we need some metrics to track it.
+	defer func() {
+		c.Debugf("Batch of %d took: %s", size, time.Since(start))
+	}()
+
+	msgsCollector := newSqsMessagesCollector(c.collectConfig, c.Entry, func(ctx context.Context, errC chan error) {
+		err := trace.NewAggregateFromChannel(errC, ctx)
+		if err != nil {
+			c.Entry.WithError(err).Error("Following errors happen during receiving/converting sqs message")
+		}
+	})
+
+	// eventsChan is used for communication between [fromSQS] and [writeToS3].
+	eventsChan := msgsCollector.getEventsChan()
+
+	readSQSCtx, readCancel := context.WithTimeout(ctx, c.batchMaxInterval)
+	defer readCancel()
+
+	go func() {
+		msgsCollector.fromSQS(readSQSCtx)
+	}()
+	var err error
+	size, err = c.writeToS3(ctx, eventsChan)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+	// TODO(tobiaszheller): delete messages from queue in next PR.
+}
+
+// sqsMessagesCollector is responsible for collecting messages from SQS and
+// writing to on channel.
+type sqsMessagesCollector struct {
+	log           *log.Entry
+	config        sqsCollectConfig
+	errHandlingFn func(ctx context.Context, errC chan error)
+	eventsChan    chan eventAndAckID
+}
+
+// newSqsMessagesCollector returns message collector.
+// Collector sends collected messages from SQS on events channel.
+func newSqsMessagesCollector(cfg sqsCollectConfig, log *log.Entry, errHandlingFn func(ctx context.Context, errC chan error)) *sqsMessagesCollector {
+	return &sqsMessagesCollector{
+		log:           log,
+		config:        cfg,
+		errHandlingFn: errHandlingFn,
+		eventsChan:    make(chan eventAndAckID, cfg.batchMaxItems),
+	}
+}
+
+// getEventsChan returns channel which can be used to read messages from SQS.
+// When collector finishes, channel will be closed.
+func (s *sqsMessagesCollector) getEventsChan() <-chan eventAndAckID {
+	return s.eventsChan
+}
+
+// fromSQS receives messages from SQS and sends it on eventsC channel.
+// It runs until context is canceled (via timeout) or when maxItems is reached.
+// MaxItems is soft limit and can happen that it will return more items then MaxItems.
+func (s *sqsMessagesCollector) fromSQS(ctx context.Context) {
+	// Errors should be immediately process by error handling loop, so 10 size
+	// should be enough to not cause blocking.
+	errorsC := make(chan error, 10)
+	defer close(errorsC)
+
+	// errhandle loop for receiving single event errors.
+	go func() {
+		s.errHandlingFn(ctx, errorsC)
+	}()
```

Comment:
I have used it to aggregate errors from whole batch and just print it once, using https://github.com/gravitational/teleport/blob/cb5ed0c46645ba69f418d31911ca60495301714c/lib/events/athena/consumer.go#L166-L169

Reply:
i am just equally afraid logging every error because if something goes wrong, we will spam logs with multiple errors. 
do you think we should aggregate those errors by error types and display count with it? and if there are <10 errors, we can display all of them?

Reply:
https://github.com/gravitational/teleport/pull/24038/commits/226f329b53975c973a37e28c60a0dac762f866b2
I have reworked it a little bit, if errors are < 10, it prints every message, if > 10, it prints only first 10 and additional emssage how many errors there were.